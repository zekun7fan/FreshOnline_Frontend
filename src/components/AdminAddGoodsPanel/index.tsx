import {Button, Form, Input, InputNumber, Modal, Select, TreeSelect} from 'antd';
import React from 'react';
import {Option} from "antd/es/mentions";

interface AdminAddGoodsPanelProps {
    visible: boolean
}



function AdminAddGoodsPanel({visible}: AdminAddGoodsPanelProps) {
    return (
            <div>
                <Modal
                    title={"ADD GOODS"}
                    visible={visible}
                    footer={[null, null]}
                    onCancel={cancel}
                >
                    <Form
                        name="goods_info"
                        onFinish={submit}
                        ref={formRef}
                    >

                        <Form.Item
                            name="name"
                            label="Name"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Input/>
                        </Form.Item>
                        <Form.Item
                            name="type"
                            label="Type"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Select>
                                <Option value={0}>By quantity</Option>
                                <Option value={1}>By weight</Option>
                            </Select>
                        </Form.Item>
                        <Form.Item
                            name={"price"}
                            label="Price"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <InputNumber
                                min="0"
                                max="100000000"
                                step="0.01"
                                stringMode
                            />
                        </Form.Item>
                        <Form.Item
                            name="storage"
                            label="Storage"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <InputNumber
                                min="0"
                                max="100000000"
                                step="0.01"
                                stringMode
                            />
                        </Form.Item>
                        <Form.Item
                            name="description"
                            label="Description">
                            <Input.TextArea/>
                        </Form.Item>
                        <Form.Item
                            name="onsale"
                            label="Onsale"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Select>
                                <Option value={0}>not onsale</Option>
                                <Option value={1}>onsale</Option>
                            </Select>
                        </Form.Item>
                        <Form.Item
                            name="sale_price"
                            label="Sale price"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <InputNumber
                                min="0"
                                max="100000000"
                                step="0.01"
                                stringMode
                            />
                        </Form.Item>
                        <Form.Item
                            name="brand"
                            label="Brand"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Input/>
                        </Form.Item>
                        <Form.Item
                            name="category_id"
                            label="Category"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <TreeSelect
                                showSearch
                                style={{width: '100%'}}
                                dropdownStyle={{maxHeight: 400, overflow: 'auto'}}
                                placeholder="Please select"
                                allowClear
                                onSelect={this.onSelect}
                            >
                                {renderTreeNode(category_tree, true)}
                            </TreeSelect>
                        </Form.Item>
                        <Form.Item
                            name="is_new"
                            label="Isnew"
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Select>
                                <Option value={0}>not new</Option>
                                <Option value={1}>new</Option>
                            </Select>
                        </Form.Item>


                        <Form.Item wrapperCol={{...this.layout.wrapperCol, offset: 8}}>
                            <Button type="primary" htmlType="submit">
                                Submit
                            </Button>
                        </Form.Item>
                    </Form>
                    <Button type="primary" onClick={this.cancel}>Cancel</Button>
                </Modal>
            </div>

    );
}


export default AdminAddGoodsPanel;