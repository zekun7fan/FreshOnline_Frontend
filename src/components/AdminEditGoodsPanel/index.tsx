import React from 'react';
import PropTypes from 'prop-types';


function AdminEditGoodsPanel() {
    return (
        <div></div>
    );
}

export default AdminEditGoodsPanel;