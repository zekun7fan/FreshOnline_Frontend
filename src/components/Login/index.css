#login_div {
    text-align: center;
    margin: auto;
    width: 60%;
}
#login_form {
    text-align: center;
    margin: auto;
}