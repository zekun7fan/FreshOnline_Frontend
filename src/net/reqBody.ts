


export interface ReqBody {

}

export interface Goods extends ReqBody{
    id?: number,
    name: string,
    type: number,
    price: number,
    storage: number,
    sales?: number,
    description?: string,
    onsale: number,
    sale_price?: number,
    rate?: number,
    rate_count?: number,
    brand: string,
    category_id: number,
    is_new: number,
}


export interface PaymentMethod extends ReqBody{
    id? : number,
    userId: number,
    paymentType: number,
    details: string
}

export interface Details extends ReqBody {
    cardNumber: number,
    cvvCode: number
}

export interface User extends ReqBody{
    id?: number,
    name?: string,
    password: string,
    email: string,
    type?: number,
    localtion?: string
}