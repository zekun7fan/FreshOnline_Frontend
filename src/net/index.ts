import {del, get, post, put} from "./http";
import {
    category,
    categoryTree,
    goods,
    goodsById,
    loginUrl, logoutUrl,
    paymentMethodByPid,
    paymentMethodByUserId,
    paymentMethodUrl, registerUrl
} from "./url";
import {AdminSearchParam} from "./reqParam";
import {Category, Goods, PaymentMethod, User} from "./reqBody";
import {CategoryNode} from "../utils/utils";


export function getCategoryTree() {
    return get(categoryTree);
}

export function getGoodsById(id: number) {
    return get(goodsById(id));
}

export function getGoods(params: AdminSearchParam) {
    return get(goods, params);
}

export function updateGoods(data: Goods) {
    return put(goods, {}, data)
}

export function addGoods(data: Goods) {
    return post(goods, {}, data)
}

export function deleteGoods(id: number) {
    return del(goodsById(id));
}


export function getPaymentMethodByUserId(id: number) {
    return get(paymentMethodByUserId(id))
}

export function addPaymentMethod(paymentMethod: PaymentMethod) {
    return post(paymentMethodUrl, {}, paymentMethod)

}

export function delPaymentMethodByPid(id: number) {
    return del(paymentMethodByPid(id))
}

export function login(user: User) {
    return put(loginUrl, {}, user)
}

export function register(user: User) {
    return post(registerUrl, {}, user)
}

export function delCategory(node: CategoryNode){
    return del(category, {}, node)
}

export function addCategory(node: Category) {
    return post(category, {}, node)
}

export function logout(user: User){
    return put(logoutUrl, {}, user)
}

export function updatePass(userId?:number, old_pass?:string, new_pass?:string) {
    return get('/user/update_password', { 'user_id': userId, "old_pass": old_pass, "new_pass": new_pass })
}
export function updateAddress(userId?:number, address?:string) {
    return get('/user/update_address', { 'user_id': userId, "address": address })
}

export function queryOrderByUser(userId?:number, startPo?:number) {
    return get('/user/orders', { 'user_id': userId, 'position': startPo })
}

export function queryWeeklySpecial() {
    return get('/weekly_special', {})
}

export function queryCategoryGoods(idList?:Array<number>|number) {
    return get('/random_goods', { 'catogory_id_list': idList })
}

export function queryUser(userId?:number) {
    return get('/user', { 'user_id': userId })
}