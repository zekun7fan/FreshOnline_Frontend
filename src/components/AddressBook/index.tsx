import React, {useEffect, useRef, useState} from 'react';
import {queryUser,updateAddress} from "../../net";
import { Form,Input,Divider,Col, Row,Button ,  Cascader,InputNumber,
                                               Select,
                                               Checkbox} from 'antd';
import {User} from "../../utils/javamodel";
import {getUserId} from "../../utils/user";

/**
 * @author Zetian Huang
 */
interface addressLines{
    text:string,
    editing: boolean,
    row:number,
    alive:boolean
}
function AddressBook() {

    const [data, setData] = useState<Array<addressLines>>([])
    const [user, setUser] = useState<User>({})
    const [content, setContent] = useState<Array<JSX.Element>>([])

    const componentDidMount = async ()=> {
        const userId = getUserId()
        if(userId==null) return
        else{
            const resp = await queryUser(userId)
            const res = resp.data
            if (res.code === 0){
                const user:User = res.data
                setUser(user)
                let location =new String(user.location)
                var array = location.split(";")
                var lindex=-1
                var arrays = array.map(
                    (ele)=>{lindex=lindex+1; return {text:ele,editing:false,row:lindex,alive:true}})
                setData(arrays)
            }
         }
    }

    useEffect(() => {
        componentDidMount()
        return () => {
        };
    }, [])

    useEffect(() => {
        getElements()
        return () => {
        };
    }, [data])


    const ApplyChange= (row:number)=>{
        data[row].editing=false
        pushUpdate()
        getElements()
    }

    const pushUpdate = async ()=>{
        var location=""
        data.map((ele)=>{location=location+( ele.alive? (ele.text+";"):"")})
        location=location.slice(0,-1)
        const resp = await updateAddress(user.id,location)
        const res = resp.data
        if (res.code !== 0){
            alert("Update fail!")
        }
    }
    const EditLocation=(row:number)=>{
        data[row].editing=true
        getElements()
    }

    const deleteLocation=(row:number)=>{
        data[row].alive=false
        pushUpdate()
        getElements()
    }

    const addLine=()=>{
        const id=data[data.length-1].row+1
        data.push({text:"",editing:true,row:id,alive:true})
        getElements()
    }
    const textChange=(id: number,obj: { target: { value: any; }; })=>{
        data[id].text=obj.target.value
    }
    const getElements = () =>{
        var lines = data.map( (ele)=>{
            if(ele.editing&&ele.alive)
            return (
                <Row key = {ele.row} gutter={16}>
                    <Col span={8}>
                        <Input defaultValue={ele.text} onChange={(obj)=>{textChange(ele.row,obj)}}/>
                    </Col>
                    <Col span={8}>
                        <Button type="link" onClick={()=>{ApplyChange(ele.row)}}>Apply</Button>
                    </Col>
                </Row>)
            else if(ele.alive) return (
                <Row key = {ele.row} gutter={16}>
                    <Col span={8}>
                        <h2>{ele.text}</h2>
                    </Col>
                    <Col span={8}>
                        <Button type="link" onClick={()=>{EditLocation(ele.row)}} >Edit</Button>
                        <Button type="link" onClick={()=>{deleteLocation(ele.row)}} >Delete</Button>
                    </Col>
                </Row>)
            else return (<div/>)
        })
        lines = lines.concat(<Button onClick={addLine}>Add</Button>)
        setContent(lines)
    }
   
    return (<div>
        {content}
    </div>);
}
export default AddressBook;
