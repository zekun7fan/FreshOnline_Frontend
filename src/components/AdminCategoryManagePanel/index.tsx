import React, {ChangeEvent, useEffect, useRef, useState} from 'react';
import {Button, Input, InputRef, message, Modal, TreeSelect} from "antd";
import {CategoryNode, getSubCategoryTree, renderTreeNode} from "../../utils/utils";
import {shallowEqual, useDispatch, useSelector} from "react-redux";
import {RootState} from "../../redux/reducers";
import {addCategory, delCategory, getCategoryTree} from "../../net";
import {Resp} from "../../net/resp";
import {update_category} from "../../redux/actions/category_tree";
import {Category} from "../../net/reqBody";


function AdminCategoryManagePanel(){

    const dispatch = useDispatch();

    const ref = useRef<InputRef>(null)

    const [visible, setVisible] = useState<boolean>(false)

    const [newCategoryName, SetNewCategoryName] = useState<string>('')

    const [selectedNode, setSelectedNode] = useState<CategoryNode>({
        id: 0,
        level: 0,
        name: 'root',
        children: []
    })

    const categoryTree = useSelector((state: RootState) => {
        return state.category_tree;
    }, shallowEqual)

    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
        SetNewCategoryName(e.target.value)
    };

    const onAddCategory = () => {
        if (selectedNode.level === 3) {
            message.warn("you can not add node more than depth 3")
            return
        }
        setVisible(true)

    }

    const onCancelAdd = () => {
        setVisible(false)
        SetNewCategoryName('')
        ref.current!.input!.value = ''
    }

    const onConfirmAdd = async () => {
        const newNode: Category = {
            name: newCategoryName,
            parentId: selectedNode.id,
            level: selectedNode.level + 1
        }
        const raw = await addCategory(newNode)
        const resp: Resp = raw.data
        if (resp.code === 0){
            reloadCategory().catch()
        }
        resetSelectedNode()
        message.info(resp.msg)
        setVisible(false)
        console.log("input=,", ref)
        ref.current!.input!.value = ''
    }

    const onDelCategory = async () => {
        const id = selectedNode.id;
        if (id === 0){
            message.warn("please select a category to delete")
            return
        }
        const node: CategoryNode = getSubCategoryTree(id, categoryTree)
        const raw = await delCategory(node)
        const resp: Resp = raw.data
        if (resp.code === 0){
            reloadCategory().catch()
        }
        message.info(resp.msg)
        resetSelectedNode()
    };

    const reloadCategory = async () => {
        const raw = await getCategoryTree()
        const resp: Resp = raw.data
        if (resp.code === 0){
            const tree = resp.data as CategoryNode[]
            dispatch(update_category(tree))
        }
    };

    const resetSelectedNode = () => {
        setSelectedNode({
            id: 0,
            level: 0,
            name: 'root',
            children: []
        })
    };

    const onSelect = (_: any, node: any) => {
        setSelectedNode({
            id: node.value,
            name: node.title,
            level: node.level,
            children: []
        })
    }

    const onClear = () => {
        resetSelectedNode()
    }


    const queryTree = async () => {
        const raw = await getCategoryTree()
        console.log("raw=",raw)
        const resp: Resp = raw.data
        if (resp.code === 0){
            const tree: CategoryNode[] = resp.data as CategoryNode[];
            dispatch(update_category(tree))
        }
    }

    useEffect(() => {
        queryTree().catch()
    },[])

    return (
            <div>
                <TreeSelect
                    style={{width: '100%'}}
                    dropdownStyle={{maxHeight: 400, overflow: 'auto'}}
                    placeholder="Please select"
                    allowClear
                    onSelect={onSelect}
                    onClear={onClear}
                >
                    {renderTreeNode(categoryTree, false)}
                </TreeSelect>
                <Button type="primary" size={'large'} onClick={onAddCategory}>
                    Add new category
                </Button>
                <Button type="primary" size={'large'} onClick={onDelCategory}>
                    Delete category
                </Button>
                <Modal title="ADD NEW CATEGORY" visible={visible} onOk={onConfirmAdd} onCancel={onCancelAdd}>
                    <p>you will create new category under {selectedNode.name} category</p>
                    new category name: <Input ref={ref} onChange={onChange}/>
                </Modal>
            </div>

    );
};


export default AdminCategoryManagePanel;