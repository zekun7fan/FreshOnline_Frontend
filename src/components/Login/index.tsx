import React from 'react';
import {Button, Col, Form, Input, message, Row} from 'antd';
import {User} from "../../net/reqBody";
import {login} from "../../net";
import {LoginedUserInfo, Resp} from "../../net/resp";
import {useNavigate} from 'react-router';
import {getUserType, user_id_key, user_name_key, user_token_key, user_type_key, UserType} from "../../utils/user";
import "./index.css"


export function Login() {
    let navigate = useNavigate();

    const onFinish = async (user: User) => {
        const raw = await login(user)
        const resp: Resp = raw.data
        if (resp.code === 0){
            const info: LoginedUserInfo = resp.data as LoginedUserInfo
            localStorage.setItem(user_id_key, String(info.id))
            localStorage.setItem(user_name_key, info.name)
            localStorage.setItem(user_type_key, String(info.type))
            localStorage.setItem(user_token_key, info.token)
            //huang added code here
            const type = getUserType()
            if (type === UserType.CUSTOMER){
                navigate(`/home`);
            }else{
                navigate(`/admin`);
            }

        }
        message.info(resp.msg)
    };

    return { localStorage, onFinish }
}

export default function LoginUI() {

    const { onFinish } = Login()

    return (
        <div id="login_div">
            <Row justify="center" align="middle" style={{minHeight: '30vh'}}>
                <Col span={24}>
                    <h2>Login</h2>
                </Col>

                <Form
                    name="user"
                    labelCol={{ span: 8 }}
                    wrapperCol={{ span: 16 }}
                    initialValues={{ remember: true }}
                    onFinish={onFinish}
                    autoComplete="off"
                    id="login_form"
                >
                    <Form.Item
                        label="Email"
                        name="email"
                        rules={[{ required: true, type: 'email' ,message: 'Please input your email!' }]}
                    >
                        <Input allowClear/>
                    </Form.Item>

                    <Form.Item
                        label="Password"
                        name="password"
                        rules={[{ required: true, message: 'Please input your password!' }]}
                    >
                        <Input.Password allowClear/>
                    </Form.Item>

                    <Form.Item wrapperCol={{ offset: 0, span: 24 }}>
                        <Button type="primary" htmlType="submit">
                            LOGIN
                        </Button>
                    </Form.Item>


                </Form>


            </Row>
        </div>
    );
}