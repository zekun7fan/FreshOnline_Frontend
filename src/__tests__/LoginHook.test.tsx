import {renderHook, act, cleanup} from '@testing-library/react-hooks';

import LoginUI from "../components/Login";
import {BrowserRouter} from "react-router-dom";
import {Login} from "../components/Login";
import {User} from "../net/reqBody";
import {jest} from "@jest/globals";
import axios from "axios";


jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

afterEach(() => {
    cleanup().then()
})

test('render login using hook', async () =>{
    const wrapper = ({ children }:{ children: any }) => (
        <BrowserRouter>
            {children}
        </BrowserRouter>
    )
    const { result } = renderHook(() => Login(), { wrapper })
    // expect(result.current).toMatchSnapshot()
    const { localStorage, onFinish } = result.current

    mockedAxios.put.mockResolvedValueOnce({data: { code:1, msg:"hello", data: {}}})
    const user: User = {}
    await act(() => onFinish(user))
    expect(typeof localStorage).toBe('object')
    expect(localStorage.length).toEqual(0)

    await cleanup()

    mockedAxios.put.mockResolvedValueOnce({data: { code:0, msg:"hello", data: {}}})
    await act(() => onFinish(user))
    expect(typeof localStorage).toBe('object')
    expect(localStorage.length).toEqual(4)
})

test('render loginUI using hook', () =>{
    const wrapper = ({ children }:{ children: any }) => (
        <BrowserRouter>
            {children}
        </BrowserRouter>
    )
    const { result } = renderHook(() => LoginUI(), { wrapper })
    expect(result.current).toMatchSnapshot()
})

afterEach(cleanup)