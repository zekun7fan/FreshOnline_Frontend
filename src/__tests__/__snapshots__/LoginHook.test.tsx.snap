// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render loginUI using hook 1`] = `
<div
  id="login_div"
>
  <Row
    align="middle"
    justify="center"
    style={
      Object {
        "minHeight": "30vh",
      }
    }
  >
    <Col
      span={24}
    >
      <h2>
        Login
      </h2>
    </Col>
    <ForwardRef(InternalForm)
      autoComplete="off"
      id="login_form"
      initialValues={
        Object {
          "remember": true,
        }
      }
      labelCol={
        Object {
          "span": 8,
        }
      }
      name="user"
      onFinish={[Function]}
      wrapperCol={
        Object {
          "span": 16,
        }
      }
    >
      <FormItem
        label="Email"
        name="email"
        rules={
          Array [
            Object {
              "message": "Please input your email!",
              "required": true,
              "type": "email",
            },
          ]
        }
      >
        <ForwardRef
          allowClear={true}
        />
      </FormItem>
      <FormItem
        label="Password"
        name="password"
        rules={
          Array [
            Object {
              "message": "Please input your password!",
              "required": true,
            },
          ]
        }
      >
        <Password
          action="click"
          allowClear={true}
          iconRender={[Function]}
          visibilityToggle={true}
        />
      </FormItem>
    </ForwardRef(InternalForm)>
    <Col
      span={24}
    >
      <Button
        htmlType="submit"
        type="primary"
      >
        LOGIN
      </Button>
    </Col>
  </Row>
</div>
`;
